<font size="6"><center><b>Life Style Data</b></center></font>

<center><img src = "../website/static/image.png" width=500px heigh=500px></center>

---

# Introduction

## Dataset Source

The Life Style dataset used in this study was sourced from Kaggle [(Essa, 2025)](https://www.kaggle.com/datasets/jockeroika/life-style-data/data). It contains fitness participant information created for recommendation systems and fitness applications, and was later expanded synthetically from 973 to 20,000 rows while maintaining logical consistency between features.

## Dataset Details

This dataset has a total of 54 features and 20000 observations.

|	Column Name	|	Description	|
| ----------- | ----------- |
|	Age	|	Age of the participant (in years).	|
|	Gender	|	Biological gender (Male/Female).	|
|	Weight (kg)	|	Weight of the individual in kilograms.	|
|	Height (m)	|	Height of the individual in meters.	|
|	Max_BPM	|	Maximum heart rate recorded during a workout session.	|
|	Avg_BPM	|	Average heart rate maintained during the session.	|
|	Resting_BPM	|	Resting heart rate before starting the workout.	|
|	Session_Duration (hours)	|	Duration of the workout session in hours.	|
|	Calories_Burned	|	Total calories burned during the session.	|
|	Workout_Type	|	Type of workout performed (e.g., Strength, HIIT, Cardio).	|
|	Fat_Percentage	|	Body fat percentage of the individual.	|
|	Water_Intake (liters)	|	Average daily water consumption in liters.	|
|	Workout_Frequency (days/week)	|	Number of workout days per week.	|
|	Experience_Level	|	Fitness experience level (1=Beginner, 2=Intermediate, 3=Advanced).	|
|	BMI	|	Body Mass Index, a measure of body fat based on height and weight.	|
|	Daily meals frequency	|	Number of meals consumed daily.	|
|	Physical exercise	|	Indicates the type or frequency of physical activity.	|
|	Carbs	|	Daily carbohydrate intake (grams).	|
|	Proteins	|	Daily protein intake (grams).	|
|	Fats	|	Daily fat intake (grams).	|
|	Calories	|	Total daily calorie intake from food.	|
|	meal_name	|	Name of the meal (e.g., Breakfast, Lunch, Dinner).	|
|	meal_type	|	Type of meal (e.g., Snack, Main, Beverage).	|
|	diet_type	|	Type of diet followed (e.g., Keto, Vegan, Balanced).	|
|	sugar_g	|	Sugar content in grams per meal.	|
|	sodium_mg	|	Sodium content in milligrams per meal.	|
|	cholesterol_mg	|	Cholesterol content in milligrams per meal.	|
|	serving_size_g	|	Portion size of the meal in grams.	|
|	cooking_method	|	Cooking method used (e.g., Boiled, Fried, Grilled).	|
|	prep_time_min	|	Preparation time in minutes.	|
|	cook_time_min	|	Cooking time in minutes.	|
|	rating	|	Meal or workout rating (typically 1–5 scale).	|
|	Name of Exercise	|	Name of the exercise performed.	|
|	Sets	|	Number of sets completed in the exercise.	|
|	Reps	|	Number of repetitions per set.	|
|	Benefit	|	Description of the exercise’s physical benefit.	|
|	Burns Calories (per 30 min)	|	Estimated calories burned in 30 minutes of that exercise.	|
|	Target Muscle Group	|	Main muscle group targeted by the exercise.	|
|	Equipment Needed	|	Equipment required to perform the exercise.	|
|	Difficulty Level	|	Exercise difficulty level (Beginner, Intermediate, Advanced).	|
|	Body Part	|	Primary body part involved (e.g., Arms, Legs, Chest).	|
|	Type of Muscle	|	Type of muscle engaged (e.g., Upper, Core, Grip Strength).	|
|	Workout	|	Specific workout or exercise name.	|

**Observation:**
- The dataset description mentions a column called `is_healthy`, intended as a boolean indicator (True/False) of whether a meal or workout is healthy. However, this column wasn't present in the dataset.
- In the dataset description (About Dataset), the author doesn't mention 12 columns that are present in the dataset, including: `BMI_calc`, `cal_from_macros`, `pct_carbs`, `protein_per_kg`, `pct_HRR`, `pct_maxHR`, `cal_balance`, `lean_mass_kg`, `expected_burn`, `Burns Calories (per 30 min)_bc`, `Burns_Calories_Bin`. It appears that only the final version of the dataset was uploaded, after the author had expanded and transformed it. Therefore, the original dataset containing 973 rows and 44 columns, and the intermediate expanded version, are not accessible.

# Importing libraries/packages for project

In [1]:
import pandas as pd
import numpy as np
from pandas.api.types import CategoricalDtype

# Importing data

In [2]:
# https://www.kaggle.com/datasets/jockeroika/life-style-data/data
data = '../data collection/Life Style Data.csv'
dataset = pd.read_csv(data)

# Exploratory Data Analysis

## Dataset Preview

In [3]:
dataset.head()

Age  Gender  Weight (kg)  Height (m)  Max_BPM  Avg_BPM  Resting_BPM  \
0  34.91    Male        65.27        1.62   188.58   157.65        69.05   
1  23.37  Female        56.41        1.55   179.43   131.75        73.18   
2  33.20  Female        58.98        1.67   175.04   123.95        54.96   
3  38.69  Female        93.78        1.70   191.21   155.10        50.07   
4  45.09    Male        52.42        1.88   193.58   152.88        70.84   

   Session_Duration (hours)  Calories_Burned Workout_Type  ...  \
0                      1.00          1080.90     Strength  ...   
1                      1.37          1809.91         HIIT  ...   
2                      0.91           802.26       Cardio  ...   
3                      1.10          1450.79         HIIT  ...   
4                      1.08          1166.40     Strength  ...   

   cal_from_macros  pct_carbs  protein_per_kg   pct_HRR  pct_maxHR  \
0          2139.59   0.500432        1.624789  0.741237   0.835985   
1          1711.65   0.500850        1.514093  0.551247   0.734270   
2          1965.92   0.500610        1.663445  0.574534   0.708124   
3          1627.28   0.499533        0.862017  0.744155   0.811150   
4          2659.23   0.500581        2.538153  0.668405   0.789751   

   cal_balance  lean_mass_kg  expected_burn  Burns Calories (per 30 min)_bc  \
0       725.10     47.777394       685.1600                    7.260425e+19   
1      -232.91     40.809803       978.6184                    1.020506e+20   
2       805.74     44.635580       654.5266                    1.079607e+20   
3      1206.21     63.007432       773.6300                    8.987921e+19   
4       303.60     43.347504       711.4176                    5.264685e+19   

   Burns_Calories_Bin  
0              Medium  
1                High  
2                High  
3                High  
4                 Low  

[5 rows x 54 columns]

## Dataset Information

In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 54 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Age                             20000 non-null  float64
 1   Gender                          20000 non-null  object 
 2   Weight (kg)                     20000 non-null  float64
 3   Height (m)                      20000 non-null  float64
 4   Max_BPM                         20000 non-null  float64
 5   Avg_BPM                         20000 non-null  float64
 6   Resting_BPM                     20000 non-null  float64
 7   Session_Duration (hours)        20000 non-null  float64
 8   Calories_Burned                 20000 non-null  float64
 9   Workout_Type                    20000 non-null  object 
 10  Fat_Percentage                  20000 non-null  float64
 11  Water_Intake (liters)           20000 non-null  float64
 12  Workout_Frequency (days/week)   

The dataset contains only two variable types: float and object (string).

## Descriptive Statistics

In [5]:
dataset.describe().T

count          mean           std  \
Age                             20000.0  3.885145e+01  1.211458e+01   
Weight (kg)                     20000.0  7.389883e+01  2.117301e+01   
Height (m)                      20000.0  1.723094e+00  1.270328e-01   
Max_BPM                         20000.0  1.798897e+02  1.151081e+01   
Avg_BPM                         20000.0  1.437043e+02  1.426769e+01   
Resting_BPM                     20000.0  6.219581e+01  7.289351e+00   
Session_Duration (hours)        20000.0  1.259446e+00  3.413362e-01   
Calories_Burned                 20000.0  1.280110e+03  5.022290e+02   
Fat_Percentage                  20000.0  2.610131e+01  4.996160e+00   
Water_Intake (liters)           20000.0  2.627486e+00  6.047240e-01   
Workout_Frequency (days/week)   20000.0  3.318629e+00  9.111979e-01   
Experience_Level                20000.0  1.809176e+00  7.362036e-01   
BMI                             20000.0  2.492168e+01  6.701126e+00   
Daily meals frequency           20000.0  2.864804e+00  6.366486e-01   
Physical exercise               20000.0  4.525030e-01  9.866276e-01   
Carbs                           20000.0  2.497786e+02  5.510862e+01   
Proteins                        20000.0  9.991829e+01  2.204367e+01   
Fats                            20000.0  6.661217e+01  1.469928e+01   
Calories                        20000.0  2.024418e+03  5.418915e+02   
sugar_g                         20000.0  2.453104e+01  1.444610e+01   
sodium_mg                       20000.0  1.258140e+03  7.166190e+02   
cholesterol_mg                  20000.0  1.484656e+02  8.738282e+01   
serving_size_g                  20000.0  3.027195e+02  1.157119e+02   
prep_time_min                   20000.0  3.351745e+01  1.652687e+01   
cook_time_min                   20000.0  6.150216e+01  3.357975e+01   
rating                          20000.0  2.989303e+00  1.157692e+00   
Sets                            20000.0  4.425059e+00  5.795959e-01   
Reps                            20000.0  1.942732e+01  3.679707e+00   
Burns Calories (per 30 min)     20000.0  3.440465e+02  3.213878e+01   
BMI_calc                        20000.0  2.492165e+01  6.701144e+00   
cal_from_macros                 20000.0  1.998297e+03  4.408484e+02   
pct_carbs                       20000.0  4.999830e-01  1.455490e-03   
protein_per_kg                  20000.0  1.460142e+00  5.189455e-01   
pct_HRR                         20000.0  6.990053e-01  1.448804e-01   
pct_maxHR                       20000.0  8.023050e-01  9.661268e-02   
cal_balance                     20000.0  7.443087e+02  7.209466e+02   
lean_mass_kg                    20000.0  5.378638e+01  1.249874e+01   
expected_burn                   20000.0  8.663523e+02  2.503171e+02   
Burns Calories (per 30 min)_bc  20000.0  8.631802e+19  3.197579e+19   

                                         min           25%           50%  \
Age                             1.800000e+01  2.817000e+01  3.986500e+01   
Weight (kg)                     3.918000e+01  5.816000e+01  7.000000e+01   
Height (m)                      1.490000e+00  1.620000e+00  1.710000e+00   
Max_BPM                         1.593100e+02  1.700575e+02  1.801400e+02   
Avg_BPM                         1.190700e+02  1.312200e+02  1.429900e+02   
Resting_BPM                     4.949000e+01  5.596000e+01  6.220000e+01   
Session_Duration (hours)        4.900000e-01  1.050000e+00  1.270000e+00   
Calories_Burned                 3.231100e+02  9.108000e+02  1.231450e+03   
Fat_Percentage                  1.133313e+01  2.238781e+01  2.582250e+01   
Water_Intake (liters)           1.460000e+00  2.170000e+00  2.610000e+00   
Workout_Frequency (days/week)   1.940000e+00  2.980000e+00  3.010000e+00   
Experience_Level                1.000000e+00  1.010000e+00  1.990000e+00   
BMI                             1.204000e+01  2.009750e+01  2.412000e+01   
Daily meals frequency           1.950000e+00  2.020000e+00  3.000000e+00   
Physical exercise              -7.000000e-02 -1.00000

*Key insights:*
- The average age of participants is approximately 38.85 years (SD = 12.1 years), with a minimum age of 18 and a maximum age of 59
- The average weight is approximately 73.9 kg (SD 21.1 kg)
- The average session duration is approximately 1.26 hours (SD 0.34 hours)
- The average calories burned is approximately 1280 (SD 502)
- The median daily meals frequency is 3 and the maximum is 4
- Both the average and the median number of sets completed per exercise are approximately 4 (SD 0.5)
- The average cooking time is approximately 61 min (SD 33.5 min)

## Data Cleaning

Steps:
- Duplicate & Missing Value Check
- Remove Redundant Features
- Data Consistency Validation
- Create Derived Variables
- Transform & Convert Types
- Standardize column names

In [6]:
# Display all the columns in the dataset
dataset.columns

Index(['Age', 'Gender', 'Weight (kg)', 'Height (m)', 'Max_BPM', 'Avg_BPM',
       'Resting_BPM', 'Session_Duration (hours)', 'Calories_Burned',
       'Workout_Type', 'Fat_Percentage', 'Water_Intake (liters)',
       'Workout_Frequency (days/week)', 'Experience_Level', 'BMI',
       'Daily meals frequency', 'Physical exercise', 'Carbs', 'Proteins',
       'Fats', 'Calories', 'meal_name', 'meal_type', 'diet_type', 'sugar_g',
       'sodium_mg', 'cholesterol_mg', 'serving_size_g', 'cooking_method',
       'prep_time_min', 'cook_time_min', 'rating', 'Name of Exercise', 'Sets',
       'Reps', 'Benefit', 'Burns Calories (per 30 min)', 'Target Muscle Group',
       'Equipment Needed', 'Difficulty Level', 'Body Part', 'Type of Muscle',
       'Workout', 'BMI_calc', 'cal_from_macros', 'pct_carbs', 'protein_per_kg',
       'pct_HRR', 'pct_maxHR', 'cal_balance', 'lean_mass_kg', 'expected_burn',
       'Burns Calories (per 30 min)_bc', 'Burns_Calories_Bin'],
      dtype='object')

### Check for Duplicate values

In [7]:
dataset.duplicated().value_counts()

False    20000
Name: count, dtype: int64

### Check for Null/Missing Values

In [8]:
dataset.isna().sum()

Age                               0
Gender                            0
Weight (kg)                       0
Height (m)                        0
Max_BPM                           0
Avg_BPM                           0
Resting_BPM                       0
Session_Duration (hours)          0
Calories_Burned                   0
Workout_Type                      0
Fat_Percentage                    0
Water_Intake (liters)             0
Workout_Frequency (days/week)     0
Experience_Level                  0
BMI                               0
Daily meals frequency             0
Physical exercise                 0
Carbs                             0
Proteins                          0
Fats                              0
Calories                          0
meal_name                         0
meal_type                         0
diet_type                         0
sugar_g                           0
sodium_mg                         0
cholesterol_mg                    0
serving_size_g              

### Drop irrelevant features

- The variable `burns calories (per 30 min)_bc` is very similar to the `burns calories (per 30 min)` feature, thus it is considered redundant for the model and removed.
- The variable `pct_carbs` is just the carbohydrate percentage, which is very similar to the `Carbs` variable, thus it is also considered redundant and has been removed.
- The variable `protein_per_kg` represents daily protein intake, but unlike the `Proteins` variable, it is measured in kilograms rather than grams. Therefore, it is considered redundant and has been removed.
- The variable `meal_name` is expected to provide information about the name of the meal (e.g., Breakfast, Lunch, Dinner), but it contains only a single value, "Other". The expected values actually appear in the variable `meal_type`. Therefore, `meal_name` is not necessary, as it does not provide any meaningful or significant information.
- The variable `cal_from_macros` is not described by the author. Since its meaning is unclear and there are already several calorie-related variables, it will be removed.
- The variable `pct_HRR` is not described by the author. It appears to represent a percentage of Heart Rate Recovery (HRR), but since its exact meaning is unclear and there is already a related `Avg_BPM` variable, it will be removed.
- The variable `pct_maxHR` is not described by the author. It appears to represent the maximum percentage of Heart Rate Recovery (HRR), but since its exact meaning is unclear and there is already a related `Max_BPM` variable, it will be removed.
- The variable `cal_balance` is not described by the author. Since its meaning is unclear and there are already several calorie-related variables, it will be removed.

In [9]:
dataset['meal_name'].unique()

array(['Other'], dtype=object)

In [10]:
dataset['meal_type'].unique()

array(['Lunch', 'Breakfast', 'Snack', 'Dinner'], dtype=object)

In [11]:
# Drop irrelevant/repeated columns/features
dataset = dataset.drop(columns=["Burns Calories (per 30 min)_bc", "pct_carbs", "protein_per_kg", "meal_name", "cal_from_macros", "pct_HRR", "pct_maxHR", "cal_balance"])

### Checking data consistency

While it was being checked whether the variable descriptions matched the data types and values, it was noticed that the variable `Physical_exercise`, which is supposed to indicate the type or frequency of physical activity, contains negative values. This is inconsistent, because frequency cannot be negative, and such an interpretation does not make sense in this context, participants cannot “undo” exercise.

Given the meaning of `Physical exercise`, negative values are not logical in this context. Since these values represent about 45.95% of the data, and because the source or nature of the error is unknown (for example, whether the negatives indicate a comparison with past frequency or are simply data entry mistakes), the decision was made to drop this column. Keeping it as it is would not provide meaningful information for the analysis.

In [12]:
dataset["Physical exercise"].describe()

count    20000.000000
mean         0.452503
std          0.986628
min         -0.070000
25%         -0.010000
50%          0.010000
75%          0.040000
max          4.050000
Name: Physical exercise, dtype: float64

In [13]:
len(dataset[dataset["Physical exercise"] <= 0])

9190

In [14]:
print(f"The negative values in Physical exercise represent {len(dataset[dataset["Physical exercise"] <= 0])/len(dataset)*100:.4}% of the data")

The negative values in Physical exercise represent 45.95% of the data


In the variable `sugar_g`, which represents sugar content in grams per meal, the same issue is present: there are negative values. A meal cannot have negative sugar content, since sugar is something consumed, not produced, by a meal.

Unlike the issue with `Physical_exercise`, the error in `sugar_g` affects only 0.35% of the data, so removing the affected rows would have minimal impact. However, because this variable is not of interest for the analysis, the entire column will be dropped.

In [15]:
dataset['sugar_g'].describe()

count    20000.000000
mean        24.531042
std         14.446105
min         -0.670000
25%         11.990000
50%         24.060000
75%         37.490000
max         50.510000
Name: sugar_g, dtype: float64

In [16]:
len(dataset[dataset["sugar_g"] <= 0])

70

In [17]:
print(f"The negative values in sugar_g represent {len(dataset[dataset["sugar_g"] <= 0])/len(dataset)*100:.4}% of the data")

The negative values in sugar_g represent 0.35% of the data


And the problem of negative values it's also present in the variable `cholesterol_mg`, which represents cholesterol content in milligrams per meal. A meal cannot contain negative cholesterol content in grams, as cholesterol is a substance found in food (from animal products) and is not something that can be subtracted.

And similarly as `sugar_g`, this error in `cholesterol_mg` represents a small percente of the data (only 0.55%), so we could simply delete the affected rows. However, because this variable is not of interest for the analysis, the entire column will be dropped.

In [18]:
dataset['cholesterol_mg'].describe()

count    20000.000000
mean       148.465602
std         87.382817
min         -3.890000
25%         71.970000
50%        149.795000
75%        221.840000
max        303.990000
Name: cholesterol_mg, dtype: float64

In [19]:
len(dataset[dataset["cholesterol_mg"] <= 0])

110

In [20]:
print(f"The negative values in cholesterol_mg represent {len(dataset[dataset["cholesterol_mg"] <= 0])/len(dataset)*100:.4}% of the data")

The negative values in cholesterol_mg represent 0.55% of the data


In [21]:
# Drop inconsistent columns
dataset = dataset.drop(columns=["Physical exercise", "sugar_g", "cholesterol_mg"])

### Create derived columns

It was noticed that the dataset’s author mentioned the variable `Age` represents the participant’s age in years. However, the data includes decimal values, which suggests that months are also included. Therefore, if a new variable called `Age_year` is created, it will represent the participant’s age rounded to whole years only.

An `Age_group` column wasn't created because the ages range from a minimum of 18 to a maximum of 59, which generally corresponds to adults (or perhaps young adults and adults). Since the difference is not substantial enough to have a significant analytical impact, grouping by age would not add meaningful insight at this stage.

In [22]:
dataset['Age'].describe()

count    20000.000000
mean        38.851454
std         12.114580
min         18.000000
25%         28.170000
50%         39.865000
75%         49.630000
max         59.670000
Name: Age, dtype: float64

In [23]:
dataset['Age_year'] = np.floor(dataset['Age']).astype("Int64")

### Transforming columns

It was noticed that the dataset’s author stated that the variable `Experience_Level` should represent the participant’s fitness experience level (1 = Beginner, 2 = Intermediate, 3 = Advanced). However, the data doesn't contain these integer values (1, 2, 3) or their corresponding category labels (Beginner, Intermediate, Advanced), instead, it displays float numbers.

To address this, the variable will be converted into categorical labels rather than keeping numeric codes. This approach will also make it easier to compare `Experience_Level` with other variables, such as `Difficulty_Level`, which use similar categorical descriptors.

In [24]:
dataset['Experience_Level'].describe()

count    20000.000000
mean         1.809176
std          0.736204
min          1.000000
25%          1.010000
50%          1.990000
75%          2.020000
max          3.050000
Name: Experience_Level, dtype: float64

In [25]:
dataset['Experience_Level'] = round(dataset['Experience_Level'])

In [26]:
dataset['Experience_Level']

0        2.0
1        2.0
2        1.0
3        2.0
4        2.0
        ... 
19995    1.0
19996    3.0
19997    2.0
19998    1.0
19999    2.0
Name: Experience_Level, Length: 20000, dtype: float64

In [27]:
# Define a mapping dictionary
category_map = {1: 'Beginner', 2: 'Intermediate', 3: 'Advanced'}

# Replace the numbers with their corresponding category names in the same column
dataset['Experience_Level'] = dataset['Experience_Level'].replace(category_map)

In [28]:
dataset['Experience_Level'].unique()

array(['Intermediate', 'Beginner', 'Advanced'], dtype=object)

### Data type conversions

It was also noticed that the dataset’s author stated that:
- The variable `Daily_meals_frequency` should represent the number of meals consumed per day.
- The variable `Sets` should represent the number of sets completed in the exercise.
- The variable `Reps` should represent the number of repetitions per set.

However, the data doesn't contain integer values; instead, it displays float numbers. Therefore, the variables will be converted into integer numbers.

In [29]:
#Before
dataset['Daily meals frequency'].unique()

array([2.99, 3.01, 1.99, 3.  , 2.02, 3.98, 2.  , 1.98, 3.02, 2.01, 4.02,
       3.03, 2.04, 4.01, 2.98, 3.99, 4.  , 4.03, 1.97, 2.97, 2.03, 3.97,
       1.96, 3.04, 3.96, 2.96, 2.95, 4.04, 3.05, 1.95, 3.95])

In [30]:
#Before
dataset['Sets'].unique()

array([4.99, 4.01, 5.  , 4.  , 5.01, 4.97, 3.99, 4.02, 4.98, 3.98, 5.02,
       5.03, 2.99, 4.03, 3.  , 3.02, 3.97, 2.98, 3.01, 4.96, 4.04, 5.05,
       2.97, 3.03, 3.96, 3.04, 2.96, 5.04, 4.05, 3.95])

In [31]:
#Before
dataset['Reps'].unique()

array([20.91, 16.15, 21.9 , 16.92, 15.01, 25.1 , 25.04, 18.95, 24.1 ,
       16.09, 17.03, 20.97, 23.95, 14.93, 15.13, 17.87, 20.9 , 16.98,
       25.07, 23.01, 22.86, 17.09, 24.02, 23.85, 14.99, 16.88, 15.03,
       22.  , 24.98, 16.01, 16.02, 18.06, 15.88, 19.04, 21.92, 10.07,
       23.09, 23.96, 23.  , 22.02, 18.  , 22.99, 21.02, 21.86, 17.95,
       21.05, 18.01, 24.13, 22.96, 14.96, 11.85, 17.11, 16.03, 20.01,
       15.15, 19.97, 23.18, 25.  , 15.91, 15.05, 11.87, 23.99, 23.06,
        8.11, 24.  , 18.99, 16.07, 25.05, 24.97, 17.97, 20.03, 17.96,
       16.91, 21.06, 14.98, 15.99, 16.05, 23.97, 15.07, 23.02, 12.03,
       18.92, 21.98, 22.98, 15.08, 24.9 , 14.95, 20.96,  9.9 , 14.86,
       21.95, 17.1 , 22.09, 20.12, 10.1 , 21.93, 18.24, 16.83, 14.76,
       23.98, 18.03, 25.12, 21.17, 22.95, 17.94, 15.  , 14.87, 17.08,
       17.  , 20.02, 15.06, 21.85, 24.04, 19.03, 20.95, 19.91, 23.89,
       21.07, 14.9 , 23.04, 16.  , 22.11, 18.88, 25.06, 25.02, 24.07,
       21.99, 22.91,

In [32]:
# Convert the float column to an integer type
dataset['Daily meals frequency'] = round(dataset['Daily meals frequency']).astype('Int64')
dataset['Sets'] = round(dataset['Sets']).astype('Int64')
dataset['Reps'] = round(dataset['Reps']).astype('Int64')

In [33]:
#After applying conversion
dataset['Daily meals frequency'].unique()

<IntegerArray>
[3, 2, 4]
Length: 3, dtype: Int64

In [34]:
#After applying conversion
dataset['Sets'].unique()

<IntegerArray>
[5, 4, 3]
Length: 3, dtype: Int64

In [35]:
#After applying conversion
dataset['Reps'].unique()

<IntegerArray>
[21, 16, 22, 17, 15, 25, 19, 24, 18, 23, 10, 12, 20, 8, 5, 30]
Length: 16, dtype: Int64

Some object-type columns will be converted into categorical variables, since categorical data is generally much more efficient for operations such as grouping, sorting, or filtering.

However, to decide whether an object-type variable should be converted into a categorical variable, the recommendations from the Pandas documentation will be followed, which suggest converting when:
- The string column contains only a few different values
- The lexical order of a variable is not the same as the logical order (e.g., “one”, “two”, “three”)
- As a signal to other Python libraries that this column should be treated as a categorical variable (e.g. to use suitable statistical methods or plot types)

The following variables are stored as strings:

In [36]:
print(dataset.select_dtypes(exclude = 'number').columns)
print(f"There're {len(dataset.select_dtypes(exclude = 'number').columns)} columns stored as a string")

Index(['Gender', 'Workout_Type', 'Experience_Level', 'meal_type', 'diet_type',
       'cooking_method', 'Name of Exercise', 'Benefit', 'Target Muscle Group',
       'Equipment Needed', 'Difficulty Level', 'Body Part', 'Type of Muscle',
       'Workout', 'Burns_Calories_Bin'],
      dtype='object')
There're 15 columns stored as a string


First, the dtype (data type) of the columns that are ordinal categories will be changed, and the correct order will be assigned to them:
- `Experience_Level`
- `Difficulty Level`
- `Burns_Calories_Bin`

In [37]:
#Initial order
print(f"Experience Level values: {dataset["Experience_Level"].unique()}")
print(f"Difficulty Level values: {dataset["Difficulty Level"].unique()}")
print(f"Burns Calories Bin values: {dataset["Burns_Calories_Bin"].unique()}")

Experience Level values: ['Intermediate' 'Beginner' 'Advanced']
Difficulty Level values: ['Advanced' 'Intermediate' 'Beginner']
Burns Calories Bin values: ['Medium' 'High' 'Low' 'Very High']


In [38]:
cat_level = CategoricalDtype(categories=["Beginner", "Intermediate", "Advanced"], ordered=True)
dataset["Experience_Level"] = dataset["Experience_Level"].astype(cat_level)

dataset["Difficulty Level"] = dataset["Difficulty Level"].astype(cat_level)

cat_Bin = CategoricalDtype(categories=["Low", "Medium", "High", "Very High"], ordered=True)
dataset["Burns_Calories_Bin"] = dataset["Burns_Calories_Bin"].astype(cat_Bin)

In [39]:
#Final order
print(f"Experience Level values: {dataset["Experience_Level"].unique()}\n")
print(f"Difficulty Level values: {dataset["Difficulty Level"].unique()}\n")
print(f"Burns_Calories_Bin values: {dataset["Burns_Calories_Bin"].unique()}")

Experience Level values: ['Intermediate', 'Beginner', 'Advanced']
Categories (3, object): ['Beginner' < 'Intermediate' < 'Advanced']

Difficulty Level values: ['Advanced', 'Intermediate', 'Beginner']
Categories (3, object): ['Beginner' < 'Intermediate' < 'Advanced']

Burns_Calories_Bin values: ['Medium', 'High', 'Low', 'Very High']
Categories (4, object): ['Low' < 'Medium' < 'High' < 'Very High']


Now, for the remaining 12 variables that are nominal categorical variables, the number of distinct values they contain will be analyzed.

In [40]:
for nunique_values in list(dataset.select_dtypes(include = 'object').columns):
    print(f"{nunique_values}: {dataset[nunique_values].nunique()}")

Gender: 2
Workout_Type: 4
meal_type: 4
diet_type: 6
cooking_method: 7
Name of Exercise: 55
Benefit: 49
Target Muscle Group: 36
Equipment Needed: 20
Body Part: 7
Type of Muscle: 13
Workout: 53


Based on the number of distinct values, there're 6 variables with more than 10 distinctive values (`Name of Exercise`, `Benefit`, `Target Muscle Group`, `Equipment Needed`, `Type of Muscle`, and `Workout`). Therefore, these will remain as object data type.

The remaining variables will be converted to categorical data types:
- `Gender`
- `Workout_Type`
- `meal_type`
- `diet_type`
- `cooking_method`
- `Body Part`

In [41]:
dataset["Gender"] = dataset["Gender"].astype('category')
dataset["Workout_Type"] = dataset["Workout_Type"].astype('category')
dataset["meal_type"] = dataset["meal_type"].astype('category')
dataset["diet_type"] = dataset["diet_type"].astype('category')
dataset["cooking_method"] = dataset["cooking_method"].astype('category')
dataset["Body Part"] = dataset["Body Part"].astype('category')

In [42]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 44 columns):
 #   Column                         Non-Null Count  Dtype   
---  ------                         --------------  -----   
 0   Age                            20000 non-null  float64 
 1   Gender                         20000 non-null  category
 2   Weight (kg)                    20000 non-null  float64 
 3   Height (m)                     20000 non-null  float64 
 4   Max_BPM                        20000 non-null  float64 
 5   Avg_BPM                        20000 non-null  float64 
 6   Resting_BPM                    20000 non-null  float64 
 7   Session_Duration (hours)       20000 non-null  float64 
 8   Calories_Burned                20000 non-null  float64 
 9   Workout_Type                   20000 non-null  category
 10  Fat_Percentage                 20000 non-null  float64 
 11  Water_Intake (liters)          20000 non-null  float64 
 12  Workout_Frequency (days/week)  2

Originally, the dataset contained dtypes are: float64(39), object(15)

Finally, After cleaning and preprocessing, the dtypes are: Int64(4), category(9), float64(25), object(6)

### Standardize column names

Some of the columns are not labeled properly, which will be problematic when modelling. All column names will be converted to lower case and any spaces will be replaced with underscores for consistency. Any white spaces before or after the column names will also be removed.

In [43]:
dataset.columns

Index(['Age', 'Gender', 'Weight (kg)', 'Height (m)', 'Max_BPM', 'Avg_BPM',
       'Resting_BPM', 'Session_Duration (hours)', 'Calories_Burned',
       'Workout_Type', 'Fat_Percentage', 'Water_Intake (liters)',
       'Workout_Frequency (days/week)', 'Experience_Level', 'BMI',
       'Daily meals frequency', 'Carbs', 'Proteins', 'Fats', 'Calories',
       'meal_type', 'diet_type', 'sodium_mg', 'serving_size_g',
       'cooking_method', 'prep_time_min', 'cook_time_min', 'rating',
       'Name of Exercise', 'Sets', 'Reps', 'Benefit',
       'Burns Calories (per 30 min)', 'Target Muscle Group',
       'Equipment Needed', 'Difficulty Level', 'Body Part', 'Type of Muscle',
       'Workout', 'BMI_calc', 'lean_mass_kg', 'expected_burn',
       'Burns_Calories_Bin', 'Age_year'],
      dtype='object')

In [44]:
# assign new names to the columns
columns_mapping = {
    'Weight (kg)': 'weight_kg',
    'Height (m)': 'height_m',
    'Session_Duration (hours)': 'session_duration_h',
    'Water_Intake (liters)': 'water_intake_l',
    'Workout_Frequency (days/week)': 'workout_frequency',
    'Daily meals frequency': 'daily_meals_frequency',
    'Name of Exercise': 'name_of_exercise',
    'Burns Calories (per 30 min)': 'burns_calories_30m',
    'Target Muscle Group': 'target_muscle_group',
    'Equipment Needed': 'equipment_needed',
    'Difficulty Level': 'difficulty_level',
    'Body Part': 'body_part', 
    'Type of Muscle': 'type_of_muscle',
    'Calories': 'calorie_intake'
}
dataset = dataset.rename(columns=columns_mapping)

In [45]:
# make column names lower case and also remove any white spaces
# before & after the column names using the strip() function
dataset.columns = dataset.columns.str.lower().str.strip()

Final names of the columns

In [46]:
dataset.columns

Index(['age', 'gender', 'weight_kg', 'height_m', 'max_bpm', 'avg_bpm',
       'resting_bpm', 'session_duration_h', 'calories_burned', 'workout_type',
       'fat_percentage', 'water_intake_l', 'workout_frequency',
       'experience_level', 'bmi', 'daily_meals_frequency', 'carbs', 'proteins',
       'fats', 'calorie_intake', 'meal_type', 'diet_type', 'sodium_mg',
       'serving_size_g', 'cooking_method', 'prep_time_min', 'cook_time_min',
       'rating', 'name_of_exercise', 'sets', 'reps', 'benefit',
       'burns_calories_30m', 'target_muscle_group', 'equipment_needed',
       'difficulty_level', 'body_part', 'type_of_muscle', 'workout',
       'bmi_calc', 'lean_mass_kg', 'expected_burn', 'burns_calories_bin',
       'age_year'],
      dtype='object')

## Final Dataset

|	Name	|	Variable Type	| Data Type	| Description	|
| ----------- | ----------- | ----------- | ----------- |
|	age	|	Numerical continuous	|	Float	|	Age of the participant (in years and months).	|
|	age_year	|	Discrete numeric	|	Integer	|	Age of the participant (in years).	|
|	gender	|	Nominal categorical	|	Category	|	Biological gender (Male/Female).	|
|	weight_kg	|	Numerical continuous	|	Float	|	Weight of the individual in kilograms.	|
|	height_m	|	Numerical continuous	|	Float	|	Height of the individual in meters.	|
|	max_bpm	|	Numerical continuous	|	Float	|	Maximum heart rate recorded during a workout session.	|
|	avg_bpm	|	Numerical continuous	|	Float	|	Average heart rate maintained during the session.	|
|	resting_bpm	|	Numerical continuous	|	Float	|	Resting heart rate before starting the workout.	|
|	 session_duration_h	|	Numerical continuous	|	Float	|	Duration of the workout session in hours.	|
|	calories_burned	|	Numerical continuous	|	Float	|	Total calories burned during the session.	|
|	workout_type	|	Nominal categorical	|	Category	|	Type of workout performed (e.g., Strength, HIIT, Cardio).	|
|	fat_percentage	|	Numerical continuous	|	Float	|	Body fat percentage of the individual.	|
|	water_intake_l	|	Numerical continuous	|	Float	|	Average daily water consumption in liters.	|
|	workout_frequency	|	Numerical continuous	|	Float	|	Number of workout days per week.	|
|	experience_level	|	Categorical ordinal	|	Category	|	Fitness experience level (1=Beginner, 2=Intermediate, 3=Advanced).	|
|	bmi	|	Numerical continuous	|	Float	|	Body Mass Index, a measure of body fat based on height and weight.	|
|	daily_meals_frequency	|	Discrete numeric	|	Integer	|	Number of meals consumed daily.	|
|	carbs	|	Numerical continuous	|	Float	|	Daily carbohydrate intake (grams).	|
|	proteins	|	Numerical continuous	|	Float	|	Daily protein intake (grams).	|
|	fats	|	Numerical continuous	|	Float	|	Daily fat intake (grams).	|
|	calorie_intake	|	Numerical continuous	|	Float	|	Total daily calorie intake from food.	|
|	meal_type	|	Nominal categorical	|	Category	|	Name of the meal (e.g., Breakfast, Lunch, Dinner).	|
|	diet_type	|	Nominal categorical	|	Category	|	Type of diet followed (e.g., Keto, Vegan, Balanced).	|
|	sodium_mg	|	Numerical continuous	|	Float	|	Sodium content in milligrams per meal.	|
|	serving_size_g	|	Numerical continuous	|	Float	|	Portion size of the meal in grams.	|
|	cooking_method	|	Nominal categorical	|	Category	|	Cooking method used (e.g., Boiled, Fried, Grilled).	|
|	prep_time_min	|	Numerical continuous	|	Float	|	Preparation time in minutes.	|
|	cook_time_min	|	Numerical continuous	|	Float	|	Cooking time in minutes.	|
|	rating	|	Numerical continuous	|	Float	|	Meal or workout rating (typically 1–5 scale).	|
|	name_of_exercise	|	Nominal categorical	|	Object (String)	|	Name of the exercise performed.	|
|	sets	|	Discrete numeric	|	Integer	|	Number of sets completed in the exercise.	|
|	reps	|	Discrete numeric	|	Integer	|	Number of repetitions per set.	|
|	benefit	|	Nominal categorical	|	Object (String)	|	Description of the exercise’s physical benefit.	|
|	burns_calories_30m	|	Numerical continuous	|	Float	|	Estimated calories burned in 30 minutes of that exercise.	|
|	target_muscle_group	|	Nominal categorical	|	Object (String)	|	Main muscle group targeted by the exercise.	|
|	equipment_needed	|	Nominal categorical	|	Object (String)	|	Equipment required to perform the exercise.	|
|	difficulty_level	|	Categorical ordinal	|	Category	|	Exercise difficulty level (Beginner, Intermediate, Advanced).	|
|	body_part	|	Nominal categorical	|	Category	|	Primary body part involved (e.g., Arms, Legs, Chest).	|
|	type_of_muscle	|	Nominal categorical	|	Object (String)	|	Type of muscle engaged (e.g., Upper, Core, Grip Strength).	|
|	workout	|	Nominal categorical	|	Object (String)	|	Specific workout or exercise name.	|
|	bmi_calc	|	Numerical continuous	|	Float	|	Body mass index (BMI)	|
|	lean_mass_kg	|	Numerical continuous	|	Float	|	Lean body mass (LBM) in kg is the weight of all the components in your body except for fat, which includes your muscles, organs, bones, and water	|
|	expected_burn	|	Numerical continuous	|	Float	|	Calories expected to be burned as a result of exercise/food	|
|	burns_calories_bin	|	Categorical ordinal	|	Category	|	Categorizes workout calories burned (e.g. Medium, High, Low, Very High)	|

In [47]:
dataset.to_csv("../data collection/new_Life_Style_Data.csv", index=False)
